include/tvm/runtime/container.h

```diff
@@ -189,6 +189,759 @@ class InplaceArrayBase {
+  static constexpr const char* _type_key = "Array";
+  TVM_DECLARE_FINAL_OBJECT_INFO(ArrayNode, Object);
+
+ private:
+  /*! \return Size of initialized memory, used by InplaceArrayBase. */
+  size_t GetSize() const { return this->size_; }
+
+  /*! \return begin mutable iterator */
+  ObjectRef* MutableBegin() const {
+    return static_cast<ObjectRef*>(InplaceArrayBase::AddressOf(0));
+  }
+
+  /*! \return end mutable iterator */
+  ObjectRef* MutableEnd() const { return MutableBegin() + size_; }
+
+  /*!
+   * \brief Create an ArrayNode with the given capacity.
+   * \param n Required capacity
+   * \return Ref-counted ArrayNode requested
+   */
+  static ObjectPtr<ArrayNode> Empty(int64_t n = kInitSize) {
+    CHECK_GE(n, 0);
+    ObjectPtr<ArrayNode> p = make_inplace_array_object<ArrayNode, ObjectRef>(n);
+    p->capacity_ = n;
+    p->size_ = 0;
+    return p;
+  }
+
+  /*!
+   * \brief Assign the elements starting idx from [first, last)
+   * \param idx The starting point
+   * \param first Begin of iterator
+   * \param last End of iterator
+   * \tparam IterType The type of iterator
+   * \return Self
+   */
+  template <typename IterType>
+  ArrayNode* AssignRange(int64_t idx, IterType first, IterType last) {
+    ObjectRef* itr = MutableBegin() + idx;
+    for (; first != last; ++first) {
+      ObjectRef ref = *first;
+      new (itr++) ObjectRef(std::move(ref));
+    }
+    return this;
+  }
+
+  /*!
+   * \brief Move elements from right to left, requires src > dst
+   * \param dst Destination
+   * \param src Source
+   * \param numel Number of elements to be moved
+   * \return Self
+   */
+  ArrayNode* MoveElementsLeft(int64_t dst, int64_t src, int64_t numel) {
+    ObjectRef* from = MutableBegin() + src;
+    ObjectRef* to = MutableBegin() + dst;
+    for (int64_t i = 0; i < numel; ++i) {
+      *to++ = std::move(*from++);
+    }
+    return this;
+  }
+
+  /*!
+   * \brief Move elements from left to rigth, requires src < dst
+   * \param dst Destination
+   * \param src Source
+   * \param numel Number of elements to be moved
+   * \return Self
+   */
+  ArrayNode* MoveElementsRight(int64_t dst, int64_t src, int64_t numel) {
+    ObjectRef* from = MutableBegin() + (src + numel);
+    ObjectRef* to = MutableBegin() + (dst + numel);
+    for (int64_t i = 0; i < numel; ++i) {
+      *--to = std::move(*--from);
+    }
+    return this;
+  }
+
+  /*!
+   * \brief Enlarge the size of the array
+   * \param delta Size enlarged, should be positive
+   * \param val Default value
+   * \return Self
+   */
+  ArrayNode* Enlarge(int64_t delta, const ObjectRef& val = ObjectRef(nullptr)) {
+    ObjectRef* itr = MutableEnd();
+    while (delta-- > 0) {
+      new (itr++) ObjectRef(val);
+      ++size_;
+    }
+    return this;
+  }
+
+  /*!
+   * \brief Shrinks the size of the array
+   * \param delta Size shrinked, should be positive
+   * \return Self
+   */
+  ArrayNode* Shrink(int64_t delta) {
```

Comment:
If we prefer pass in delta, let us do `ShrinkBy`, alternatively, to make it looks like resize, pass in the target value, and use `ShrinkTo` (perhaps later is more consistent with STL)

Reply:
I think we could do `ShrinkBy` because it avoids some repetitive arithmetics. There is `std::vector::shrink_to_fit` in STL, which is mostly used for releasing memory in `[size, capacity)`, and thus I think it serves different purposes thought.